src/snowflake/snowpark/_internal/ast_utils.py

```diff
@@ -386,53 +386,34 @@ def build_udtf_apply(
     build_fn_apply_args(ast, *args, **kwargs)
 
 
-def build_session_table_fn_apply(
+def build_call_table_function_apply(
     ast: proto.Expr,
-    name: Union[str, Iterable[str]],
+    name: str,
     *args: Tuple[Union[proto.Expr, Any]],
     **kwargs: Dict[str, Union[proto.Expr, Any]],
 ) -> None:
-    """
-    Creates AST encoding for ApplyExpr(SessionTableFn(<name>), List(<args...>), Map(<kwargs...>)) for session table functions.
-    Args:
-        ast: Expr node to fill.
-        name: Name of the session table function to call.
-        *args: Positional arguments to pass to function.
-        **kwargs: Keyword arguments to pass to function.
-    """
     expr = with_src_position(ast.apply_expr)
-    _set_fn_name(name, expr.fn.session_table_fn)
-    with_src_position(expr.fn.session_table_fn)
+    _set_fn_name(name, expr.fn.call_table_function_expr)
+    with_src_position(expr.fn.call_table_function_expr)
     build_fn_apply_args(ast, *args, **kwargs)
 
 
-def build_table_fn_apply(
+def build_indirect_table_fn_apply(
     ast: proto.Expr,
-    name: Union[str, Iterable[str], None],
-    *args: Tuple[Union[proto.Expr, Any]],
-    **kwargs: Dict[str, Union[proto.Expr, Any]],
+    func: Union[str, List[str], "snowflake.snowpark.table_function.TableFunctionCall"],
+    *func_arguments: ColumnOrName,
+    **func_named_arguments: ColumnOrName,
 ) -> None:
-    """
-    Creates AST encoding for ApplyExpr(TableFn(<name>), List(<args...>), Map(<kwargs...>)) for table functions.
-    Args:
-        ast: Expr node to fill.
-        name: Name of the table function to call. The name can be None and is ignored for table function calls of type SessionTableFn.
-        *args: Positional arguments to pass to function.
-        **kwargs: Keyword arguments to pass to function.
-
-    Requires that ast.apply_expr.fn.table_fn.call_type is set to a valid TableFnCallType.
-    """
     expr = with_src_position(ast.apply_expr)
-    assert (
-        ast.apply_expr.fn.table_fn.call_type.WhichOneof("variant") is not None
-    ), f"Explicitly set the call type before calling this function {str(ast.apply_expr.fn.table_fn)}"
-    if not expr.fn.table_fn.call_type.table_fn_call_type__session_table_fn:
-        assert (
-            name is not None
-        ), f"Table function name must be provided {str(ast.apply_expr.fn.table_fn)}"
-        _set_fn_name(name, expr.fn.table_fn)
-    with_src_position(expr.fn.table_fn)
-    build_fn_apply_args(ast, *args, **kwargs)
+    if hasattr(func, "_ast_stmt"):
+        stmt = func._ast_stmt
+        fn_expr = expr.fn.indirect_table_fn_call_ref
+        fn_expr.id.bitfield1 = stmt.var_id.bitfield1
+    else:
+        fn_expr = expr.fn.indirect_table_fn_name_call
+        _set_fn_name(func, fn_expr)
+    with_src_position(fn_expr)
```

Comment:
the `with_src_position` should be moved outside of this helper function, i.e.
```
expr = with_src_position(...)
build_indirect_table_fn_apply(expr, ...)
```
This helper func should only be concerned with building up the expr (without symbol/src extract)

Reply:
I think this pattern matches what `build_builtin_fn_apply` and other sibling functions do. We should establish a uniform convention and follow it everywhere.

Also, if a convention requires all callers to do the exact same work everywhere, maybe the work does belong inside the callee.